hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/WritableState.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.spi.state;
+
+import java.util.Optional;
+import java.util.Set;
+import javax.annotation.Nonnull;
+
+/**
+ * A {@link ReadableState} who's values can be mutated.
+ *
+ * @param <K> The key, which must be of the appropriate kind depending on whether it is stored in
+ *     memory, or on disk.
+ * @param <V> The value, which must be of the appropriate kind depending on whether it is stored in
+ *     memory, or on disk.
+ */
+public interface WritableState<K, V> extends ReadableState<K, V> {
+
+    /**
+     * Gets the value associated with the given key in a <strong>READ-WRITE</strong> way. The
+     * returned {@link Optional} will be empty if the key does not exist in the store. If the value
+     * did exist, but the expiration time has been exceeded, then the value will be unset in the
+     * {@link Optional}.
+     *
+     * @param key The key. Cannot be null, otherwise an exception is thrown.
+     * @return A non-null optional. It may be empty if there is no value for this associated key.
+     * @throws NullPointerException if the key is null.
+     */
+    @Nonnull
+    Optional<V> getForModify(@Nonnull K key);
```

Comment:
Some of Danno's recent [Besu performance improvements](https://github.com/hashgraph/hedera-services/pull/4120) seem to consist primarily of removing `Optional` wrappers; might the `Map` precedent of returning `null` for missing keys give us some reason to trust people's intuitions without the `Optional`?

Reply:
This is a good question. I prefer optional, but it does come at a cost, always having to wrap everything. Maybe it isn't worth it. I wonder if we should leave it in for the time being until we get some performance numbers?

Reply:
I ended up removing Optional. With @NonNull and @Nullable, we get good static analysis which basically gives us what Optional was giving us, albeit not as strongly as a compiler